### PR TITLE
[Workflow] Ensure asynchronous execution of "workflow.run" & fix TODOs

### PR DESCRIPTION
**Pull request by suquark**
## Why are these changes needed?

Currently, `workflow.run` is blocking, this is not acceptable because when we implement virtual actors later, we would reuse some code path of `workflow.run`, and we would not expect virtual actors been blocked. We address this by creating a remove function in `workflow_access.py`. This module would also contain other code that can access workflow status in the future.

This PR also address several issues we did not fix last time:

1. Implement readable default workflow ID and step ID. The ID now contains the name of the entrypoint workflow step (for workflow ID) and normalized function names (for step ID).
2. Improved logging. Log both resuming and workflow running, the storage we are using, and the completion of a workflow.
3. Exception message. When a workflow fails, raise an exception including the workflow ID.
4. Replace `workflow_root_dir` with "storage_url", since we will support general storage later (e.g, S3)
5. Move code related to workflow step execution to another module. This makes code more clear.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/experimental/workflow/api.py [hunk 1]
```diff
@@ -30,18 +26,16 @@ def step(func) -> WorkflowStepFunction:
 
 # TODO(suquark): Raise an error when calling run() on an existing workflow.
 # Maybe we can also add a run_or_resume() call.
-# TODO(suquark): Rename "workflow_root_dir" to "storage_url".
 
 
-def run(entry_workflow: Workflow, workflow_root_dir=None,
+def run(entry_workflow: Workflow, storage_url=None,
```
Comment (ericl) [inline python/ray/experimental/workflow/api.py:31]: Could we just make this `storage: Storage = None` instead?

Comment (ericl) [inline python/ray/experimental/workflow/api.py:31]: Could we just make this `storage: Union[str, Storage] = None` instead? Then you can either pass in a custom storage implementation, or a s3/filesystem uri.

File: python/ray/experimental/workflow/workflow_access.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+
+class WorkflowExecutionError(Exception):
+    def __init__(self, workflow_id: str):
+        self.message = f"Workflow[id={workflow_id}] failed during execution."
+        super().__init__(self.message)
+
+
+@ray.remote
+def workflow_output_cache(workflow_id: str, output: ray.ObjectRef) -> Any:
```
Comment (ericl) [inline python/ray/experimental/workflow/workflow_access.py:16]: cache doesn't make sense as the verb here, how about `resolve_workflow_outputs` instead?

File: python/ray/experimental/workflow/api.py [hunk 3]
```diff
@@ -50,21 +44,19 @@ def run(entry_workflow: Workflow, workflow_root_dir=None,
...
-                                                    workflow_root_dir)
+        workflow_context.init_workflow_step_context(workflow_id, storage_url)
         rref = postprocess_workflow_step(entry_workflow)
         logger.info(f"Workflow job {workflow_id} started.")
-        # TODO(suquark): although we do not return the resolved object to user,
-        # the object was resolved temporarily to the driver script.
-        # We may need a helper step for storing the resolved object
-        # instead later.
-        output = resolve_object_ref(rref)[1]
+        output = workflow_output_cache.remote(workflow_id, rref)
```
Comment (ericl) [inline python/ray/experimental/workflow/api.py:59]: Can we add a test this is non-blocking?

**Comment by iycheng** [comment]
I feel good about this one. The only concern is that in resolving the task, eventually we'll get the physical result in the py layer, and if it's big, it won't be efficient and maybe even fail due to OOM in the local task. 
I'm wondering whether there is something better we can do to avoid it?

File: python/ray/experimental/workflow/api.py [hunk 4]
```diff
@@ -28,14 +29,16 @@ def step(func) -> WorkflowStepFunction:
 # Maybe we can also add a run_or_resume() call.
 
 
-def run(entry_workflow: Workflow, storage_url=None,
-        workflow_id=None) -> ray.ObjectRef:
+def run(entry_workflow: Workflow,
+        storage_or_url: Optional[Union[str, storage.Storage]] = None,
```
Comment (ericl) [inline python/ray/experimental/workflow/api.py:33]:
```suggestion
        storage: Optional[Union[str, storage.Storage]] = None,
```

File: python/ray/experimental/workflow/api.py [hunk 5]
```diff
@@ -28,14 +29,16 @@ def step(func) -> WorkflowStepFunction:
...
+        workflow_id: Optional[str] = None) -> ray.ObjectRef:
     """
     Run a workflow asynchronously.
 
     Args:
         entry_workflow: The workflow to run.
-        storage_url: The URL of an external storage used for checkpointing.
+        storage_or_url: The storage or the URL of an external storage used for
```
Comment (ericl) [inline python/ray/experimental/workflow/api.py:40]:
```suggestion
        storage: The storage or the URL of an external storage used for
```

File: python/ray/experimental/workflow/api.py [hunk 6]
```diff
@@ -65,32 +71,38 @@ def run(entry_workflow: Workflow, storage_url=None,
 # TODO(suquark): support recovery with ObjectRef inputs.
 
 
-def resume(workflow_id: str, storage_url=None) -> ray.ObjectRef:
+def resume(workflow_id: str,
+           storage_or_url: Optional[Union[str, storage.Storage]] = None
```
Comment (ericl) [inline python/ray/experimental/workflow/api.py:75]:
```suggestion
           storage: Optional[Union[str, storage.Storage]] = None
```

File: python/ray/experimental/workflow/api.py [hunk 7]
```diff
@@ -65,32 +71,38 @@ def run(entry_workflow: Workflow, storage_url=None,
...
+           ) -> ray.ObjectRef:
     """
     Resume a workflow asynchronously. This workflow maybe fail previously.
 
     Args:
         workflow_id: The ID of the workflow. The ID is used to identify
             the workflow.
-        storage_url: The path of an external storage used for
+        storage_or_url: The storage or the URL of an external storage used for
```
Comment (ericl) [inline python/ray/experimental/workflow/api.py:83]:
```suggestion
        storage: The storage or the URL of an external storage used for
```

**Review by ericl — COMMENTED** [review]
LGTM but please rename it to just storage

**Comment by ericl** [comment]
> I feel good about this one. The only concern is that in resolving the task, eventually we'll get the physical result in the py layer, and if it's big, it won't be efficient and maybe even fail due to OOM in the local task.

How about using ray.wait(fetch=False) instead?

**Comment by iycheng** [comment]
> > I feel good about this one. The only concern is that in resolving the task, eventually we'll get the physical result in the py layer, and if it's big, it won't be efficient and maybe even fail due to OOM in the local task.
> 
> How about using ray.wait(fetch=False) instead?

`fetch=False` might not be enough. The resolver function should return a object_ref where in the driver, ray.get(object_ref) will return the actual data directly. 
There are two difficulties here:
1. when returning the function, it'll always create an object_ref owned by the caller. So unless we have the object, we can't return it as an object ref directly pointing to the object.
2. Right now we don't know how deep the object ref is. The condition used here is to check whether it's object ref or not. Until dereference the object ref, we don't know the actual data. It's not very difficult to fix although, for example, we can always pass some meta together with object ref back. But 1) is still an issue here.

**Comment by ericl** [comment]
How about using multiple returns to handle this then? The second return can
be a small status object, and we can use that to decide whether there are
more nested objects to resolve.

**Comment by iycheng** [comment]
> How about using multiple returns to handle this then? The second return can be a small status object, and we can use that to decide whether there are more nested objects to resolve.

I think this sounds like a doable plan, the only issue is that the nested chain will be longer, but I think maybe it's OK.

**Comment by suquark** [comment]
I am not sure what is the solution or problem to solve according to the comments. I will discuss and solve the problem tomorrow.

**Comment by iycheng** [comment]
> I am not sure what is the solution or problem to solve according to the comments. I will discuss and solve the problem tomorrow. To me the biggest problem is `ray.wait` does not tell us if the underlying object is a physical object or object ref, unless we `ray.get` it.

What @ericl  suggest is that, we creating things like this:  
```
def f():
   r = ray.put(obj)
  return WorkflowTypeHint("REF", r)

```
Basically, the first `ray.get` will tell you the type of value behind and you can decide by the return type.

My concern is that:

```
def resolver(obj_ref):
    # resolve to the last obj_leaf
   # ray.get(obj_leaf) will return 10GB data
   return obj_leaf

# driver
r = resolver(nested_obj_ref)
a = ray.get(r) -> ObjectRef
ray.get(a) -> 10GB dat

```

**Comment by suquark** [comment]
@iycheng unfortunately a bug is preventing me from further progress: https://github.com/ray-project/ray/issues/16792

**Comment by suquark** [comment]
blocked by https://github.com/ray-project/ray/pull/16727

**Comment by suquark** [comment]
I have to rebase this PR to adopt to upstream changes.

**Comment by suquark** [comment]
@ericl @iycheng could you review the PR again? (mainly the last 2 commits)

**Comment by iycheng** [comment]
Shall we have a test for the self-dereferencer? Others look fine.

**Comment by suquark** [comment]
@iycheng actually I already implemented an almost identical test when fixing the bug

**Comment by iycheng** [comment]
> @iycheng actually I already implemented an almost identical test when fixing the bug. maybe not need to do it again

They are testing different things. Here it's testing the de-referencer you have here. And that is testing the ray bugs. If you update the way you dereference the ref here in this code, it won't be covered. Some simple tests should be enough.

**Comment by suquark** [comment]
@iycheng I add the test. The CI errors seem unrelated.

**Review by fishbone — APPROVED** [review]

**Comment by suquark** [comment]
I'll merge it since there are no more comments, and this PR blocks almost all following PRs.

**Pull request merged**